src/d3a/models/myco_matcher/__init__.py

```diff
@@ -8,13 +10,14 @@
 
 class MycoMatcher:
     def __init__(self):
-        self.bid_offer_pairs = []
         if ConstSettings.IAASettings.BID_OFFER_MATCH_TYPE == \
                 BidOfferMatchAlgoEnum.PAY_AS_BID.value:
             self.match_algorithm = PayAsBidMatcher()
         elif ConstSettings.IAASettings.BID_OFFER_MATCH_TYPE == \
                 BidOfferMatchAlgoEnum.PAY_AS_CLEAR.value:
             self.match_algorithm = PayAsClearMatcher()
+        elif is_custom_matching_enabled():
+            self.match_algorithm = ExternalMatcher()
```

Comment:
I would suggest to uniform this and the other checks (here you use a function, while the other `if` clauses don't). Maybe it would even possible to use a dictionary for the matching, like:
```python
MARKET_TO_MATCHING_ALGORITHM_MAPPING = {
    BidOfferMatchAlgoEnum.PAY_AS_CLEAR.value: PayAsClearMatcher,
    BidOfferMatchAlgoEnum.PAY_AS_BID.value: PayAsBidMatcher,
    BidOfferMatchAlgoEnum.CUSTOM.value: ExternalMatcher,
}
```
and then you could just do
```python
self.match_algorithm = MARKET_TO_MATCHING_ALGORITHM_MAPPING[ConstSettings.IAASettings.BID_OFFER_MATCH_TYPE]
```
Or something similar.

Reply:
If the logic you want to enforce is more complex, you could create a class for it (like `MatchingAlgorithmMapper.get_matching_algorithm()`)

Reply:
This can work for setting the Match Algorithm, however the `is_custom_matching_enabled` is intended as a shortcut for checking if the external matching is enabled, not only for setting the matching algorithm.

Reply:
I'm still not fully convinced. Right now, `is_external_matching_enabled()` has to be called just to decide if we're going to call some specific methods on one specific matcher, and the matcher class itself is instantiated based on this check. So I see some kind of redundancy.
I have the feeling that this could be refactored taking advantage of the standard interface for all matchers (the `BaseMatcher`). We could have one method shared by all children, and they would implement a different internal logic.

Reply:
The problem that Hannes was pointing to is that in several places in the code we wanted to check if the external matching is enabled [See here](https://github.com/gridsingularity/d3a/blob/ceb0cd569ebd369540e69c5c502f993157bd94fb/src/d3a/d3a_core/simulation.py#L341)
I can encapsulate the algorithm selection in a function if you believe this is an addition, or I can create another function which we can pass the matching type and it verifies if it is set globally, and the `is_external_matching_enabled` will then call it with the corresponding argument.
If you have something else in mind let me know

Reply:
My idea is to avoid having more small functions that we need to call multiple times across the code. Also, these functions might be populated with `if` statements, which complicates stuff. That's why I am suggesting alternative ways, more SOLID-friendly, that I believe would be more robust and easier to test.
However, you have the biggest knowledge of this code. So if you believe that my suggestion is not practical and the current implementation is a better way to proceed, I will give my LGTM 👍

Reply:
I will keep working in the same area in Test 2, will try to tackle this then